src/int.rs

```diff
@@ -0,0 +1,261 @@
+//! This module contains an 256-bit signed integer implementation.
+
+use crate::errors::ParseI256Error;
+use std::fmt;
+use std::str::{self, FromStr};
+use web3::types::U256;
+
+/// Compute the two's complement of a U256.
+fn twos_complement(u: U256) -> U256 {
+    let (twos_complement, _) = (!u).overflowing_add(U256::one());
+    twos_complement
+}
+
+/// Little-endian 256-bit signed integer.
+#[derive(Clone, Copy, Default, Eq, Hash, PartialEq)]
+pub struct I256(U256);
+
+/// Enum to represent the sign of a 256-bit signed integer.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum Sign {
+    /// Greater than or equal to zero.
+    Positive,
+    /// Less than zero.
+    Negative,
+}
+
+impl I256 {
+    /// Creates an I256 from a sign and an absolute value. Returns the value and
+    /// a bool that is true if the conversion caused an overflow.
+    fn overflowing_from_sign_and_abs(sign: Sign, abs: U256) -> (Self, bool) {
+        let value = I256(match sign {
+            Sign::Positive => abs,
+            Sign::Negative => twos_complement(abs),
+        });
+        (value, value.sign() != sign)
+    }
+
+    /// Creates an I256 from an absolute value and a negative flag. Returns
+    /// `None` if it would overflow an `I256`.
+    fn checked_from_sign_and_abs(sign: Sign, abs: U256) -> Option<Self> {
+        let (result, overflow) = I256::overflowing_from_sign_and_abs(sign, abs);
+        if overflow {
+            None
+        } else {
+            Some(result)
+        }
+    }
+
+    /// Splits a I256 into its absolute value and negative flag.
+    fn into_sign_and_abs(self) -> (Sign, U256) {
+        let sign = self.sign();
+        let abs = match sign {
+            Sign::Positive => self.0,
+            Sign::Negative => twos_complement(self.0),
+        };
+        (sign, abs)
+    }
+
+    /// Returns the sign of self.
+    fn sign(self) -> Sign {
+        let most_significant_word = (self.0).0[3];
+        match most_significant_word & (1 << 63) {
+            0 => Sign::Positive,
+            _ => Sign::Negative,
+        }
+    }
```

Comment:
I would add a test for this function. I think this is incorrect but I am not sure .`0 x ... x` should be negative, `1 0 ... 0` negative, and all other numbers positive.

Reply:
This gets implicitly tested with the formatting tests (for displaying the `-` character). From what I gather from solidity, signed negative numbers are represented as their two's complements. That means that `-I256::MAX == -(1<<255)` is `b100...0` and `-1` is `b111..1`. Either way, a negative number should always have its MSb set.

Reply:
I was confusing how two's complement is stored in my head. It's correct.